idtoken/idtoken.go

```diff
@@ -103,45 +108,72 @@ func newTokenSource(ctx context.Context, audience string, ds *internal.DialSetti
 }
 
 func tokenSourceFromBytes(ctx context.Context, data []byte, audience string, ds *internal.DialSettings) (oauth2.TokenSource, error) {
-	if err := isServiceAccount(data); err != nil {
-		return nil, err
-	}
-	cfg, err := google.JWTConfigFromJSON(data, ds.GetScopes()...)
+	allowedType, err := getAllowedType(data)
 	if err != nil {
 		return nil, err
 	}
-
-	customClaims := ds.CustomClaims
-	if customClaims == nil {
-		customClaims = make(map[string]interface{})
-	}
-	customClaims["target_audience"] = audience
-
-	cfg.PrivateClaims = customClaims
-	cfg.UseIDToken = true
-
-	ts := cfg.TokenSource(ctx)
-	tok, err := ts.Token()
-	if err != nil {
-		return nil, err
+	if allowedType == "service_account" {
+		cfg, err := google.JWTConfigFromJSON(data, ds.GetScopes()...)
+		if err != nil {
+			return nil, err
+		}
+		customClaims := ds.CustomClaims
+		if customClaims == nil {
+			customClaims = make(map[string]interface{})
+		}
+		customClaims["target_audience"] = audience
+
+		cfg.PrivateClaims = customClaims
+		cfg.UseIDToken = true
+
+		ts := cfg.TokenSource(ctx)
+		tok, err := ts.Token()
+		if err != nil {
+			return nil, err
+		}
+		return oauth2.ReuseTokenSource(tok, ts), nil
+	} else {
+		// if allowedType is "impersonated_service_account":
+		type url struct {
+			ServiceAccountImpersonationURL string `json:"service_account_impersonation_url"`
+		}
+		var accountUrl *url
+		if err := json.Unmarshal(data, &accountUrl); err != nil {
+			return nil, err
+		}
+		account := filepath.Base(accountUrl.ServiceAccountImpersonationURL)
+		account = strings.Split(account, ":")[0]
+
+		config := impersonate.IDTokenConfig{
+			Audience:        audience,
+			TargetPrincipal: account,
+			IncludeEmail:    true,
+		}
+		ts, err := impersonate.IDTokenSource(ctx, config)
+		if err != nil {
+			log.Println(err)
+		}
+		return ts, nil
 	}
-	return oauth2.ReuseTokenSource(tok, ts), nil
 }
 
-func isServiceAccount(data []byte) error {
+// isOfAllowedType returns the credentials type as a string, and an error.
+// allowed types are "service_account" and "impersonated_service_account"
+func getAllowedType(data []byte) (string, error) {
 	if len(data) == 0 {
-		return fmt.Errorf("idtoken: credential provided is 0 bytes")
+		return "", fmt.Errorf("idtoken: credential provided is 0 bytes")
 	}
 	var f struct {
 		Type string `json:"type"`
 	}
+	// if not service account return an error
 	if err := json.Unmarshal(data, &f); err != nil {
-		return err
+		return "", err
 	}
-	if f.Type != "service_account" {
-		return fmt.Errorf("idtoken: credential must be service_account, found %q", f.Type)
+	if f.Type != "service_account" && f.Type != "impersonated_service_account" {
+		return "", fmt.Errorf("idtoken: credential must be service_account or impersonated_service_account, found %q", f.Type)
```

Comment:
More types may be added over time, be less specific and maybe just say it is an unsupported type.